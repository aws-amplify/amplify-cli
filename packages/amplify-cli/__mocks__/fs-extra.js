const fsExtra = jest.genMockFromModule('fs-extra');

function copySync() {
  return {};
}

fsExtra.copySync = copySync;

module.exports = fsExtra;
