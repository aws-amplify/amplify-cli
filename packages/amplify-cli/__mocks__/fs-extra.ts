const fs: any = jest.genMockFromModule('fs-extra');

fs.readdirSync = jest.fn().mockReturnValue(['file1', 'file2']);
fs.readFileSync = jest.fn().mockReturnValue({});
fs.statSync = jest.fn().mockReturnValue({ isDirectory: () => {} });
fs.unlinkSync = jest.fn().mockReturnValue(null);
fs.copySync = jest.fn().mockReturnValue({});

module.exports = fs;
