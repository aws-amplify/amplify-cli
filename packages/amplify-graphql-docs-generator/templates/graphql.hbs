# this is an auto generated file. This will be overwritten
{{#each operations }}
  {{> renderOp }}
{{/each}}
{{#each fragments }}
  {{> renderExternalFragment }}
{{/each}}