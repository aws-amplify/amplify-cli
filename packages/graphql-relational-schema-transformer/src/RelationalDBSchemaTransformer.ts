import { Kind, ObjectTypeDefinitionNode, SchemaDefinitionNode, InputObjectTypeDefinitionNode, DocumentNode } from 'graphql';
import {
  getSingletonListTypeNode,
  getNamedType,
  getOperationFieldDefinition,
  getNonNullType,
  getInputValueDefinition,
  getTypeDefinition,
  getFieldDefinition,
  getDirectiveNode,
  getOperationTypeDefinition,
} from './RelationalDBSchemaTransformerUtils';
import { RelationalDBParsingException } from './RelationalDBParsingException';
import { IRelationalDBReader } from './IRelationalDBReader';
import { plurality, toUpper } from 'graphql-transformer-common';

/**
 * This class is used to transition all of the columns and key metadata from a table for use
 * in generating appropriate GraphQL schema structures. It will track type definitions for
 * the base table, update mutation inputs, create mutation inputs, and primary key metadata.
 */
export class TableContext {
  tableTypeDefinition: ObjectTypeDefinitionNode;
  createTypeDefinition: InputObjectTypeDefinitionNode;
  updateTypeDefinition: InputObjectTypeDefinitionNode;
  // Table primary key metadata, to help properly key queries and mutations.
  tableKeyField: string;
  tableKeyFieldType: string;
  stringFieldList: string[];
  intFieldList: string[];
  constructor(
    typeDefinition: ObjectTypeDefinitionNode,
    createDefinition: InputObjectTypeDefinitionNode,
    updateDefinition: InputObjectTypeDefinitionNode,
    primaryKeyField: string,
    primaryKeyType: string,
    stringFieldList: string[],
    intFieldList: string[],
  ) {
    this.tableTypeDefinition = typeDefinition;
    this.tableKeyField = primaryKeyField;
    this.createTypeDefinition = createDefinition;
    this.updateTypeDefinition = updateDefinition;
    this.tableKeyFieldType = primaryKeyType;
    this.stringFieldList = stringFieldList;
    this.intFieldList = intFieldList;
  }
}

/**
 * This class is used to transition all of the information needed to generate the
 * CloudFormation template. This is the class that is outputted by the SchemaTransformer and the one that
 * RelationalDBTemplateGenerator takes in for the constructor. It tracks the graphql schema document,
 * map of the primary keys for each of the types. It is also being used to track the CLI inputs needed
 * for DataSource Creation, as data source creation is apart of the cfn template generation.
 */
export class TemplateContext {
  schemaDoc: DocumentNode;
  typePrimaryKeyMap: Map<string, string>;
  typePrimaryKeyTypeMap: Map<string, string>;
  stringFieldMap: Map<string, string[]>;
  intFieldMap: Map<string, string[]>;
  secretStoreArn: string;
  rdsClusterIdentifier: string;
  databaseName: string;
  databaseSchema: string;
  region: string;

  constructor(
    schemaDoc: DocumentNode,
    typePrimaryKeyMap: Map<string, string>,
    stringFieldMap: Map<string, string[]>,
    intFieldMap: Map<string, string[]>,
    typePrimaryKeyTypeMap?: Map<string, string>,
  ) {
    this.schemaDoc = schemaDoc;
    this.typePrimaryKeyMap = typePrimaryKeyMap;
    this.stringFieldMap = stringFieldMap;
    this.intFieldMap = intFieldMap;
    this.typePrimaryKeyTypeMap = typePrimaryKeyTypeMap;
  }
}

export class RelationalDBSchemaTransformer {
  dbReader: IRelationalDBReader;
  database: string;
  improvePluralization: boolean;

  constructor(dbReader: IRelationalDBReader, database: string, improvePluralization: boolean) {
    this.dbReader = dbReader;
    this.database = database;
    this.improvePluralization = improvePluralization;
  }

  public introspectDatabaseSchema = async (): Promise<TemplateContext | null> => {
    // Get all of the tables within the provided db
    let tableNames = null;

    try {
      tableNames = await this.dbReader.listTables();
    } catch (err) {
      throw new RelationalDBParsingException(`Failed to list tables in ${this.database}`, err.stack);
    }

    // Return early if there are no tables in the database
    if (tableNames.length === 0) {
      return null;
    }

    const typeContexts = new Array();
    const types = new Array();
    const pkeyMap = new Map<string, string>();
    const pkeyTypeMap = new Map<string, string>();
    const stringFieldMap = new Map<string, string[]>();
    const intFieldMap = new Map<string, string[]>();

    for (const tableName of tableNames) {
      let type: TableContext = null;

      try {
        type = await this.dbReader.describeTable(tableName);
      } catch (err) {
        throw new RelationalDBParsingException(`Failed to describe table ${tableName}`, err.stack);
      }

      // NOTE from @mikeparisstuff. The GraphQL schema generation breaks
      // when the table does not have an explicit primary key.
      if (type.tableKeyField) {
        typeContexts.push(type);
        // Generate the 'connection' type for each table type definition
        // TODO: Determine if Connection is needed as Data API doesn't provide pagination
        // TODO: As we add different db sources, we should conditionally do this even if we don't for Aurora serverless.
        //types.push(this.getConnectionType(tableName))
        // Generate the create operation input for each table type definition
        types.push(type.createTypeDefinition);
        // Generate the default shape for the table's structure
        types.push(type.tableTypeDefinition);
        // Generate the update operation input for each table type definition
        types.push(type.updateTypeDefinition);

        // Update the field map with the new field lists for the current table
        stringFieldMap.set(tableName, type.stringFieldList);
        intFieldMap.set(tableName, type.intFieldList);
        pkeyMap.set(tableName, type.tableKeyField);
        pkeyTypeMap.set(tableName, type.tableKeyFieldType);
      } else {
        console.warn(`Skipping table ${type.tableTypeDefinition.name.value} because it does not have a single PRIMARY KEY.`);
      }
    }

    if (typeContexts.length > 0) {
      // Generate the mutations and queries based on the table structures
      types.push(this.getMutations(typeContexts));
      types.push(this.getQueries(typeContexts));
      types.push(this.getSubscriptions(typeContexts));
      types.push(this.getSchemaType());
    }

    let context = this.dbReader.hydrateTemplateContext(
      new TemplateContext({ kind: Kind.DOCUMENT, definitions: types }, pkeyMap, stringFieldMap, intFieldMap, pkeyTypeMap),
    );

    return context;
  };

  /**
   * Creates a schema type definition node, including operations for each of query, mutation, and subscriptions.
   *
   * @returns a basic schema definition node.
   */
  getSchemaType(): SchemaDefinitionNode {
    return {
      kind: Kind.SCHEMA_DEFINITION,
      directives: [],
      operationTypes: [
        getOperationTypeDefinition('query', getNamedType('Query')),
        getOperationTypeDefinition('mutation', getNamedType('Mutation')),
        getOperationTypeDefinition('subscription', getNamedType('Subscription')),
      ],
    };
  }

  /**
   * Generates the basic mutation operations, given the provided table contexts. This will
   * create a create, delete, and update operation for each table.
   *
   * @param types the table contexts from which the mutations are to be generated.
   * @returns the type definition for mutations, including a create, delete, and update for each table.
   */
  private getMutations(types: TableContext[]): ObjectTypeDefinitionNode {
    const fields = [];

    for (const typeContext of types) {
      const type = typeContext.tableTypeDefinition;
      const formattedTypeValue = toUpper(type.name.value);

      fields.push(
        getOperationFieldDefinition(
          `delete${formattedTypeValue}`,
          [getInputValueDefinition(getNonNullType(getNamedType(typeContext.tableKeyFieldType)), typeContext.tableKeyField)],
          getNamedType(`${type.name.value}`),
          null,
        ),
      );

      fields.push(
        getOperationFieldDefinition(
          `create${formattedTypeValue}`,
          [getInputValueDefinition(getNonNullType(getNamedType(`Create${formattedTypeValue}Input`)), `create${formattedTypeValue}Input`)],
          getNamedType(`${type.name.value}`),
          null,
        ),
      );

      fields.push(
        getOperationFieldDefinition(
          `update${formattedTypeValue}`,
          [getInputValueDefinition(getNonNullType(getNamedType(`Update${formattedTypeValue}Input`)), `update${formattedTypeValue}Input`)],
          getNamedType(`${type.name.value}`),
          null,
        ),
      );
    }

    return getTypeDefinition(fields, 'Mutation');
  }

  /**
   * Generates the basic subscription operations, given the provided table contexts. This will
   * create an onCreate subscription for each table.
   *
   * @param types the table contexts from which the subscriptions are to be generated.
   * @returns the type definition for subscriptions, including an onCreate for each table.
   */
  private getSubscriptions(types: TableContext[]): ObjectTypeDefinitionNode {
    const fields = [];

    for (const typeContext of types) {
      const type = typeContext.tableTypeDefinition;
      const formattedTypeValue = toUpper(type.name.value);

      fields.push(
        getOperationFieldDefinition(`onCreate${formattedTypeValue}`, [], getNamedType(`${type.name.value}`), [
          getDirectiveNode(`create${formattedTypeValue}`),
        ]),
      );
    }

    return getTypeDefinition(fields, 'Subscription');
  }

  /**
   * Generates the basic query operations, given the provided table contexts. This will
   * create a get and list operation for each table.
   *
   * @param types the table contexts from which the queries are to be generated.
   * @returns the type definition for queries, including a get and list for each table.
   */
  private getQueries(types: TableContext[]): ObjectTypeDefinitionNode {
    const fields = [];

    for (const typeContext of types) {
      const type = typeContext.tableTypeDefinition;
      const formattedTypeValue = toUpper(type.name.value);

      fields.push(
        getOperationFieldDefinition(
          `get${formattedTypeValue}`,
          [getInputValueDefinition(getNonNullType(getNamedType(typeContext.tableKeyFieldType)), typeContext.tableKeyField)],
          getNamedType(`${type.name.value}`),
          null,
        ),
      );

      // use list type node to match the ast of current schema built by graphql.parse
      const nameListType = getSingletonListTypeNode(type.name.value);

      fields.push(getOperationFieldDefinition(`list${plurality(formattedTypeValue, this.improvePluralization)}`, [], nameListType, null));
    }

    return getTypeDefinition(fields, 'Query');
  }

  /**
   * Creates a GraphQL connection type for a given GraphQL type, corresponding to a SQL table name.
   *
   * @param tableName the name of the SQL table (and GraphQL type).
   * @returns a type definition node defining the connection type for the provided type name.
   */
  getConnectionType(tableName: string): ObjectTypeDefinitionNode {
    return getTypeDefinition(
      [getFieldDefinition('items', getNamedType(`[${tableName}]`)), getFieldDefinition('nextToken', getNamedType('String'))],
      `${tableName}Connection`,
    );
  }
}
