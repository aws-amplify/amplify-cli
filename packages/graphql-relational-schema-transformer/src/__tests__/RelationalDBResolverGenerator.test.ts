import * as fs from 'fs-extra';
import { RelationalDBResolverGenerator } from '../RelationalDBResolverGenerator';
import { TemplateContext } from '../RelationalDBSchemaTransformer';
import { parse } from 'graphql';

jest.mock('fs-extra', () => ({
  writeFileSync: jest.fn(),
}));
const writeFileSync_mock = fs.writeFileSync as jest.MockedFunction<typeof fs.writeFileSync>;
afterEach(() => jest.clearAllMocks());

/**
 * Test for verifying that provided a template context, the resolver generator
 * creates the CRUDL AppSync Resolver resources.
 */
test('Test Basic CRUDL Resolver Generation', () => {
  // SETUP
  const schema = parse(`
      type Pet {
        id: String
        name: String
      }

      type Owner {
        id: String
        name: String
      }
    `);
  let simpleStringFieldMap = new Map<string, string[]>();
  let simpleIntFieldMap = new Map<string, string[]>();
  let simplePrimaryKeyMap = new Map<string, string>();
  let simplePrimaryKeyTypeMap = new Map<string, string>();

  simplePrimaryKeyMap.set('Pet', 'Id');
  simplePrimaryKeyMap.set('Owner', 'Id');
  simplePrimaryKeyTypeMap.set('Pet', 'String');
  simplePrimaryKeyTypeMap.set('Owner', 'Int');
  const context = new TemplateContext(schema, simplePrimaryKeyMap, simpleStringFieldMap, simpleIntFieldMap, simplePrimaryKeyTypeMap);
  const generator = new RelationalDBResolverGenerator(context);

  // TEST
  const resources: { [key: string]: any } = generator.createRelationalResolvers('someFilePath', true);

  // VERIFY
  expect(resources).toBeDefined();

  // Verify all CRUDL resolvers were created for the Pet Type
  expect(resources).toHaveProperty('PetCreateResolver');
  expect(resources).toHaveProperty('PetGetResolver');
  expect(resources).toHaveProperty('PetUpdateResolver');
  expect(resources).toHaveProperty('PetDeleteResolver');
  expect(resources).toHaveProperty('PetListResolver');

  // Verify for the GetResolver the elements are present
  let resolverMap = Object.keys(resources).map(key => resources[key]);
  expect(resolverMap[1]).toHaveProperty('Type');
  expect(resolverMap[1]).toHaveProperty('Properties');

  // Verify a resolver was created for the owner type as well
  expect(resources).toHaveProperty('OwnerCreateResolver');
});

test('verify generated templates', () => {
  // SETUP
  const schema = parse(`
      type Tomatoes {
        id: String
        name: String
      }
    `);
  let simpleStringFieldMap = new Map<string, string[]>();
  let simpleIntFieldMap = new Map<string, string[]>();
  let simplePrimaryKeyMap = new Map<string, string>();
  let simplePrimaryKeyTypeMap = new Map<string, string>();

  simplePrimaryKeyMap.set('Tomatoes', 'Id');
  simplePrimaryKeyTypeMap.set('Tomatoes', 'String');
  const context = new TemplateContext(schema, simplePrimaryKeyMap, simpleStringFieldMap, simpleIntFieldMap, simplePrimaryKeyTypeMap);
  const generator = new RelationalDBResolverGenerator(context);
  generator.createRelationalResolvers('testFilePath', true);
  expect(writeFileSync_mock.mock.calls.length).toBe(10);
  writeFileSync_mock.mock.calls.forEach(call => {
    expect(call.length).toBe(3);
    expect(call[0]).toMatchSnapshot();
    expect(call[1]).toMatchSnapshot();
    expect(call[2]).toBe('utf8');
  });
});

test('verify generated templates using a Int primary key', () => {
  // SETUP
  const schema = parse(`
      type Apples {
        id: Int
        name: String
      }
    `);
  let simpleStringFieldMap = new Map<string, string[]>();
  let simpleIntFieldMap = new Map<string, string[]>();
  let simplePrimaryKeyMap = new Map<string, string>();
  let simplePrimaryKeyTypeMap = new Map<string, string>();

  simplePrimaryKeyMap.set('Apples', 'Id');
  simplePrimaryKeyTypeMap.set('Apples', 'Int');
  const context = new TemplateContext(schema, simplePrimaryKeyMap, simpleStringFieldMap, simpleIntFieldMap, simplePrimaryKeyTypeMap);
  const generator = new RelationalDBResolverGenerator(context);
  generator.createRelationalResolvers('testFilePath', true);
  expect(writeFileSync_mock.mock.calls.length).toBe(10);
  writeFileSync_mock.mock.calls.forEach(call => {
    expect(call.length).toBe(3);
    expect(call[0]).toMatchSnapshot();
    expect(call[1]).toMatchSnapshot();
    expect(call[2]).toBe('utf8');
  });
});

test('verify generated templates with old pluralization', () => {
  // SETUP
  const schema = parse(`
      type Apples {
        id: Int
        name: String
      }
    `);
  let simpleStringFieldMap = new Map<string, string[]>();
  let simpleIntFieldMap = new Map<string, string[]>();
  let simplePrimaryKeyMap = new Map<string, string>();
  let simplePrimaryKeyTypeMap = new Map<string, string>();

  simplePrimaryKeyMap.set('Apples', 'Id');
  simplePrimaryKeyTypeMap.set('Apples', 'Int');
  const context = new TemplateContext(schema, simplePrimaryKeyMap, simpleStringFieldMap, simpleIntFieldMap, simplePrimaryKeyTypeMap);
  const generator = new RelationalDBResolverGenerator(context);
  generator.createRelationalResolvers('testFilePath', false);
  expect(writeFileSync_mock.mock.calls.length).toBe(10);
  writeFileSync_mock.mock.calls.forEach(call => {
    expect(call.length).toBe(3);
    expect(call[0]).toMatchSnapshot();
    expect(call[1]).toMatchSnapshot();
    expect(call[2]).toBe('utf8');
  });
});
