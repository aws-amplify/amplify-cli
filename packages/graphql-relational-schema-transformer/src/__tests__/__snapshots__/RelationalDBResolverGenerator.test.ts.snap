// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verify generated templates 1`] = `"testFilePath/Mutation.createTomatoes.req.vtl"`;

exports[`verify generated templates 2`] = `
"#set( $cols = [] )
#set( $vals = [] )
#foreach( $entry in $ctx.args.createTomatoesInput.keySet() )
  #set( $discard = $cols.add($entry) )
  #set( $discard = $vals.add(\\"'$ctx.args.createTomatoesInput[$entry]'\\") )
#end
#set( $valStr = $vals.toString().replace(\\"[\\",\\"(\\").replace(\\"]\\",\\")\\") )
#set( $colStr = $cols.toString().replace(\\"[\\",\\"(\\").replace(\\"]\\",\\")\\") )
{
  \\"version\\": \\"2018-05-29\\",
  \\"statements\\":   [\\"INSERT INTO Tomatoes $colStr VALUES $valStr\\", \\"SELECT * FROM Tomatoes WHERE Id='$ctx.args.createTomatoesInput.Id'\\"]
}"
`;

exports[`verify generated templates 3`] = `"testFilePath/Mutation.createTomatoes.res.vtl"`;

exports[`verify generated templates 4`] = `"$utils.toJson($utils.parseJson($utils.rds.toJsonString($ctx.result))[1][0])"`;

exports[`verify generated templates 5`] = `"testFilePath/Query.getTomatoes.req.vtl"`;

exports[`verify generated templates 6`] = `
"{
  \\"version\\": \\"2018-05-29\\",
  \\"statements\\":   [\\"SELECT * FROM Tomatoes WHERE Id='$ctx.args.Id'\\"]
}"
`;

exports[`verify generated templates 7`] = `"testFilePath/Query.getTomatoes.res.vtl"`;

exports[`verify generated templates 8`] = `
"#set( $output = $utils.rds.toJsonObject($ctx.result) )
#if( $output.isEmpty() )
  $util.error(\\"Invalid response from RDS DataSource. See info for the full response.\\", \\"InvalidResponse\\", {}, $output)
#end
#set( $output = $output[0] )
#if( $output.isEmpty() )
  #return
#end
$utils.toJson($output[0])"
`;

exports[`verify generated templates 9`] = `"testFilePath/Mutation.updateTomatoes.req.vtl"`;

exports[`verify generated templates 10`] = `
"#set( $updateList = {} )
#foreach( $entry in $ctx.args.updateTomatoesInput.keySet() )
  #set( $discard = $updateList.put($entry, \\"'$ctx.args.updateTomatoesInput[$entry]'\\") )
#end
#set( $update = $updateList.toString().replace(\\"{\\",\\"\\").replace(\\"}\\",\\"\\") )
{
  \\"version\\": \\"2018-05-29\\",
  \\"statements\\":   [\\"UPDATE Tomatoes SET $update WHERE Id='$ctx.args.updateTomatoesInput.Id'\\", \\"SELECT * FROM Tomatoes WHERE Id='$ctx.args.updateTomatoesInput.Id'\\"]
}"
`;

exports[`verify generated templates 11`] = `"testFilePath/Mutation.updateTomatoes.res.vtl"`;

exports[`verify generated templates 12`] = `
"#set( $output = $utils.rds.toJsonObject($ctx.result) )
#if( $output.length() < 2 )
  $util.error(\\"Invalid response from RDS DataSource. See info for the full response.\\", \\"InvalidResponse\\", {}, $output)
#end
#set( $output = $output[1] )
#if( $output.isEmpty() )
  #return
#end
$utils.toJson($output[0])"
`;

exports[`verify generated templates 13`] = `"testFilePath/Mutation.deleteTomatoes.req.vtl"`;

exports[`verify generated templates 14`] = `
"{
  \\"version\\": \\"2018-05-29\\",
  \\"statements\\":   [\\"SELECT * FROM Tomatoes WHERE Id='$ctx.args.Id'\\", \\"DELETE FROM Tomatoes WHERE Id='$ctx.args.Id'\\"]
}"
`;

exports[`verify generated templates 15`] = `"testFilePath/Mutation.deleteTomatoes.res.vtl"`;

exports[`verify generated templates 16`] = `
"#set( $output = $utils.rds.toJsonObject($ctx.result) )
#if( $output.isEmpty() )
  $util.error(\\"Invalid response from RDS DataSource. See info for the full response.\\", \\"InvalidResponse\\", {}, $output)
#end
#set( $output = $output[0] )
#if( $output.isEmpty() )
  #return
#end
$utils.toJson($output[0])"
`;

exports[`verify generated templates 17`] = `"testFilePath/Query.listTomatoes.req.vtl"`;

exports[`verify generated templates 18`] = `
"{
  \\"version\\": \\"2018-05-29\\",
  \\"statements\\":   [\\"SELECT * FROM Tomatoes\\"]
}"
`;

exports[`verify generated templates 19`] = `"testFilePath/Query.listTomatoes.res.vtl"`;

exports[`verify generated templates 20`] = `"$utils.toJson($utils.rds.toJsonObject($ctx.result)[0])"`;

exports[`verify generated templates using a Int primary key 1`] = `"testFilePath/Mutation.createApples.req.vtl"`;

exports[`verify generated templates using a Int primary key 2`] = `
"#set( $cols = [] )
#set( $vals = [] )
#foreach( $entry in $ctx.args.createApplesInput.keySet() )
  #set( $discard = $cols.add($entry) )
  #set( $discard = $vals.add(\\"'$ctx.args.createApplesInput[$entry]'\\") )
#end
#set( $valStr = $vals.toString().replace(\\"[\\",\\"(\\").replace(\\"]\\",\\")\\") )
#set( $colStr = $cols.toString().replace(\\"[\\",\\"(\\").replace(\\"]\\",\\")\\") )
{
  \\"version\\": \\"2018-05-29\\",
  \\"statements\\":   [\\"INSERT INTO Apples $colStr VALUES $valStr\\", \\"SELECT * FROM Apples WHERE Id=$ctx.args.createApplesInput.Id\\"]
}"
`;

exports[`verify generated templates using a Int primary key 3`] = `"testFilePath/Mutation.createApples.res.vtl"`;

exports[`verify generated templates using a Int primary key 4`] = `"$utils.toJson($utils.parseJson($utils.rds.toJsonString($ctx.result))[1][0])"`;

exports[`verify generated templates using a Int primary key 5`] = `"testFilePath/Query.getApples.req.vtl"`;

exports[`verify generated templates using a Int primary key 6`] = `
"{
  \\"version\\": \\"2018-05-29\\",
  \\"statements\\":   [\\"SELECT * FROM Apples WHERE Id=$ctx.args.Id\\"]
}"
`;

exports[`verify generated templates using a Int primary key 7`] = `"testFilePath/Query.getApples.res.vtl"`;

exports[`verify generated templates using a Int primary key 8`] = `
"#set( $output = $utils.rds.toJsonObject($ctx.result) )
#if( $output.isEmpty() )
  $util.error(\\"Invalid response from RDS DataSource. See info for the full response.\\", \\"InvalidResponse\\", {}, $output)
#end
#set( $output = $output[0] )
#if( $output.isEmpty() )
  #return
#end
$utils.toJson($output[0])"
`;

exports[`verify generated templates using a Int primary key 9`] = `"testFilePath/Mutation.updateApples.req.vtl"`;

exports[`verify generated templates using a Int primary key 10`] = `
"#set( $updateList = {} )
#foreach( $entry in $ctx.args.updateApplesInput.keySet() )
  #set( $discard = $updateList.put($entry, \\"'$ctx.args.updateApplesInput[$entry]'\\") )
#end
#set( $update = $updateList.toString().replace(\\"{\\",\\"\\").replace(\\"}\\",\\"\\") )
{
  \\"version\\": \\"2018-05-29\\",
  \\"statements\\":   [\\"UPDATE Apples SET $update WHERE Id=$ctx.args.updateApplesInput.Id\\", \\"SELECT * FROM Apples WHERE Id=$ctx.args.updateApplesInput.Id\\"]
}"
`;

exports[`verify generated templates using a Int primary key 11`] = `"testFilePath/Mutation.updateApples.res.vtl"`;

exports[`verify generated templates using a Int primary key 12`] = `
"#set( $output = $utils.rds.toJsonObject($ctx.result) )
#if( $output.length() < 2 )
  $util.error(\\"Invalid response from RDS DataSource. See info for the full response.\\", \\"InvalidResponse\\", {}, $output)
#end
#set( $output = $output[1] )
#if( $output.isEmpty() )
  #return
#end
$utils.toJson($output[0])"
`;

exports[`verify generated templates using a Int primary key 13`] = `"testFilePath/Mutation.deleteApples.req.vtl"`;

exports[`verify generated templates using a Int primary key 14`] = `
"{
  \\"version\\": \\"2018-05-29\\",
  \\"statements\\":   [\\"SELECT * FROM Apples WHERE Id=$ctx.args.Id\\", \\"DELETE FROM Apples WHERE Id=$ctx.args.Id\\"]
}"
`;

exports[`verify generated templates using a Int primary key 15`] = `"testFilePath/Mutation.deleteApples.res.vtl"`;

exports[`verify generated templates using a Int primary key 16`] = `
"#set( $output = $utils.rds.toJsonObject($ctx.result) )
#if( $output.isEmpty() )
  $util.error(\\"Invalid response from RDS DataSource. See info for the full response.\\", \\"InvalidResponse\\", {}, $output)
#end
#set( $output = $output[0] )
#if( $output.isEmpty() )
  #return
#end
$utils.toJson($output[0])"
`;

exports[`verify generated templates using a Int primary key 17`] = `"testFilePath/Query.listApples.req.vtl"`;

exports[`verify generated templates using a Int primary key 18`] = `
"{
  \\"version\\": \\"2018-05-29\\",
  \\"statements\\":   [\\"SELECT * FROM Apples\\"]
}"
`;

exports[`verify generated templates using a Int primary key 19`] = `"testFilePath/Query.listApples.res.vtl"`;

exports[`verify generated templates using a Int primary key 20`] = `"$utils.toJson($utils.rds.toJsonObject($ctx.result)[0])"`;

exports[`verify generated templates with old pluralization 1`] = `"testFilePath/Mutation.createApples.req.vtl"`;

exports[`verify generated templates with old pluralization 2`] = `
"#set( $cols = [] )
#set( $vals = [] )
#foreach( $entry in $ctx.args.createApplesInput.keySet() )
  #set( $discard = $cols.add($entry) )
  #set( $discard = $vals.add(\\"'$ctx.args.createApplesInput[$entry]'\\") )
#end
#set( $valStr = $vals.toString().replace(\\"[\\",\\"(\\").replace(\\"]\\",\\")\\") )
#set( $colStr = $cols.toString().replace(\\"[\\",\\"(\\").replace(\\"]\\",\\")\\") )
{
  \\"version\\": \\"2018-05-29\\",
  \\"statements\\":   [\\"INSERT INTO Apples $colStr VALUES $valStr\\", \\"SELECT * FROM Apples WHERE Id=$ctx.args.createApplesInput.Id\\"]
}"
`;

exports[`verify generated templates with old pluralization 3`] = `"testFilePath/Mutation.createApples.res.vtl"`;

exports[`verify generated templates with old pluralization 4`] = `"$utils.toJson($utils.parseJson($utils.rds.toJsonString($ctx.result))[1][0])"`;

exports[`verify generated templates with old pluralization 5`] = `"testFilePath/Query.getApples.req.vtl"`;

exports[`verify generated templates with old pluralization 6`] = `
"{
  \\"version\\": \\"2018-05-29\\",
  \\"statements\\":   [\\"SELECT * FROM Apples WHERE Id=$ctx.args.Id\\"]
}"
`;

exports[`verify generated templates with old pluralization 7`] = `"testFilePath/Query.getApples.res.vtl"`;

exports[`verify generated templates with old pluralization 8`] = `
"#set( $output = $utils.rds.toJsonObject($ctx.result) )
#if( $output.isEmpty() )
  $util.error(\\"Invalid response from RDS DataSource. See info for the full response.\\", \\"InvalidResponse\\", {}, $output)
#end
#set( $output = $output[0] )
#if( $output.isEmpty() )
  #return
#end
$utils.toJson($output[0])"
`;

exports[`verify generated templates with old pluralization 9`] = `"testFilePath/Mutation.updateApples.req.vtl"`;

exports[`verify generated templates with old pluralization 10`] = `
"#set( $updateList = {} )
#foreach( $entry in $ctx.args.updateApplesInput.keySet() )
  #set( $discard = $updateList.put($entry, \\"'$ctx.args.updateApplesInput[$entry]'\\") )
#end
#set( $update = $updateList.toString().replace(\\"{\\",\\"\\").replace(\\"}\\",\\"\\") )
{
  \\"version\\": \\"2018-05-29\\",
  \\"statements\\":   [\\"UPDATE Apples SET $update WHERE Id=$ctx.args.updateApplesInput.Id\\", \\"SELECT * FROM Apples WHERE Id=$ctx.args.updateApplesInput.Id\\"]
}"
`;

exports[`verify generated templates with old pluralization 11`] = `"testFilePath/Mutation.updateApples.res.vtl"`;

exports[`verify generated templates with old pluralization 12`] = `
"#set( $output = $utils.rds.toJsonObject($ctx.result) )
#if( $output.length() < 2 )
  $util.error(\\"Invalid response from RDS DataSource. See info for the full response.\\", \\"InvalidResponse\\", {}, $output)
#end
#set( $output = $output[1] )
#if( $output.isEmpty() )
  #return
#end
$utils.toJson($output[0])"
`;

exports[`verify generated templates with old pluralization 13`] = `"testFilePath/Mutation.deleteApples.req.vtl"`;

exports[`verify generated templates with old pluralization 14`] = `
"{
  \\"version\\": \\"2018-05-29\\",
  \\"statements\\":   [\\"SELECT * FROM Apples WHERE Id=$ctx.args.Id\\", \\"DELETE FROM Apples WHERE Id=$ctx.args.Id\\"]
}"
`;

exports[`verify generated templates with old pluralization 15`] = `"testFilePath/Mutation.deleteApples.res.vtl"`;

exports[`verify generated templates with old pluralization 16`] = `
"#set( $output = $utils.rds.toJsonObject($ctx.result) )
#if( $output.isEmpty() )
  $util.error(\\"Invalid response from RDS DataSource. See info for the full response.\\", \\"InvalidResponse\\", {}, $output)
#end
#set( $output = $output[0] )
#if( $output.isEmpty() )
  #return
#end
$utils.toJson($output[0])"
`;

exports[`verify generated templates with old pluralization 17`] = `"testFilePath/Query.listAppless.req.vtl"`;

exports[`verify generated templates with old pluralization 18`] = `
"{
  \\"version\\": \\"2018-05-29\\",
  \\"statements\\":   [\\"SELECT * FROM Apples\\"]
}"
`;

exports[`verify generated templates with old pluralization 19`] = `"testFilePath/Query.listAppless.res.vtl"`;

exports[`verify generated templates with old pluralization 20`] = `"$utils.toJson($utils.rds.toJsonObject($ctx.result)[0])"`;
